### PR TITLE
EIP-3860 Initcode Size Reference Test fixes

### PR DESCRIPTION
**Pull request by shemnon**
## PR description

Fix corner cases around initcode size checking in reference tests.

## Documentation

- [ ] I thought about documentation and added the `doc-change-required` label to this PR if
    [updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).

## Changelog

- [ ] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: evm/src/main/java/org/hyperledger/besu/evm/operation/AbstractCreateOperation.java [hunk 1]
```diff
@@ -74,6 +78,10 @@ public OperationResult execute(final MessageFrame frame, final EVM evm) {
 
       final long inputOffset = clampedToLong(frame.getStackItem(1));
       final long inputSize = clampedToLong(frame.getStackItem(2));
+      if (inputSize > maxInitcodeSize) {
+        frame.popStackItems(getStackItemsConsumed());
+        return new OperationResult(cost, ExceptionalHaltReason.CODE_TOO_LARGE);
+      }
```
Comment (siladu) [inline evm/src/main/java/org/hyperledger/besu/evm/operation/AbstractCreateOperation.java:84]: Does this warrant a test at the Operation level?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetTransactionValidator.java [hunk 2]
```diff
@@ -184,6 +189,14 @@ public ValidationResult<TransactionInvalidReason> validate(
               intrinsicGasCost, transaction.getGasLimit()));
     }
 
+    if (transaction.getTo().isEmpty() && transaction.getPayload().size() > maxInitcodeSize) {
```
Comment (siladu) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetTransactionValidator.java:192]:
```suggestion
    if (transaction.isContractCreation() && transaction.getPayload().size() > maxInitcodeSize) {
```

File: evm/src/main/java/org/hyperledger/besu/evm/operation/CreateOperation.java [hunk 3]
```diff
@@ -21,8 +21,8 @@
 
 public class CreateOperation extends AbstractCreateOperation {
 
-  public CreateOperation(final GasCalculator gasCalculator) {
-    super(0xF0, "CREATE", 3, 1, gasCalculator);
+  public CreateOperation(final GasCalculator gasCalculator, final int maxInitcodeSize) {
```
Comment (siladu) [inline evm/src/main/java/org/hyperledger/besu/evm/operation/CreateOperation.java:24]: Can you explain why we need the check at the EVM level as well as transaction validation level?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetTransactionValidator.java (hunk 2 above)
Comment (siladu) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetTransactionValidator.java:192]: Also, think this is missing a test as well?

**Review by siladu — COMMENTED** [review]
Looks pretty straight forward. Just a couple of questions a minor suggestion

File: evm/src/main/java/org/hyperledger/besu/evm/operation/CreateOperation.java (hunk 3 above)
Comment (shemnon) [reply evm/src/main/java/org/hyperledger/besu/evm/operation/CreateOperation.java:24]: Transaction validation is top level and checked before any EVM operations are executed.  Once we are in the operations we have accepted the transaction as valid and are executing it. These are two distinct different cases.  The EIP also recognizes this as rules 1&3 cover validation and recognize the different contexts.

File: evm/src/main/java/org/hyperledger/besu/evm/operation/AbstractCreateOperation.java (hunk 1 above)
Comment (shemnon) [reply evm/src/main/java/org/hyperledger/besu/evm/operation/AbstractCreateOperation.java:84]: Tests were there, but testing the wrong thing.  Fixed.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetTransactionValidator.java (hunk 2 above)
Comment (shemnon) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetTransactionValidator.java]: yep.  Added.

File: evm/src/test/java/org/hyperledger/besu/evm/operations/Create2OperationTest.java [hunk 4]
```diff
@@ -171,16 +182,124 @@ public void setUp() {
...
+      final String code,
+      final String ignoredExpectedAddress,
+      final int expectedGas) {
+    setUp(sender, salt, code);
     final OperationResult result = operation.execute(messageFrame, evm);
     assertThat(result.getHaltReason()).isNull();
     assertThat(result.getGasCost()).isEqualTo(expectedGas);
   }
+
+  @org.junit.Test
```
Comment (siladu) [inline evm/src/test/java/org/hyperledger/besu/evm/operations/Create2OperationTest.java:212]:
```suggestion
  @Test
```

**Review by siladu — APPROVED** [review]
LGTM

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetProtocolSpecs.java [hunk 5]
```diff
@@ -690,17 +686,19 @@ static ProtocolSpecBuilder shanghaiDefinition(
...
-                        CachedInvalidCodeRule.of(EvmSpecVersion.SHANGHAI)),
-                    1,
-                    SPURIOUS_DRAGON_FORCE_DELETE_WHEN_EMPTY_ADDRESSES))
+                    chainId,
+                    Set.of(
+                        TransactionType.FRONTIER,
+                        TransactionType.ACCESS_LIST,
+                        TransactionType.EIP1559),
+                    quorumCompatibilityMode,
+                    SHANGHAI_INIT_CODE_SIZE_LIMIT))
```
Comment (siladu) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetProtocolSpecs.java:701]: @shemnon I just spotted this...should it be
configContractSizeLimit.orElse(SHANGHAI_INIT_CODE_SIZE_LIMIT)?

Comment (shemnon) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetProtocolSpecs.java:701]: No.  We are limiting initcocde, Which is twice the contract size limit.  initcode and contract code have two different limits.
